File: scripts/db/generation-history.mjs
/**
 * Read-only: genererings-historik för backoffice-panelen `Generation History`.
 *
 * Slår ihop de tabeller som pipelinen redan skriver per generering:
 *   - `generation_telemetry`        (utfall per version: scaffold, model, retry,
 *                                     autofix, preflight, quality gate, preview)
 *   - latest `product_postcheck.summary` → `product_blocked` +
 *     `reported_quality_gate` (SM-017: visad grind, inte stored enum)
 *   - `engine_versions`             (version-rader: F2/F3 lifecycle, release/verify-state)
 *   - `engine_chats` + `app_projects` (chatt/projekt-kontext för läsbara etiketter)
 *   - `engine_version_error_logs`   (per-version fel/warnings, t.ex. merge:cross-file-stub)
 *   - `engine_generation_logs`      (model/tokens/duration/success per generering)
 *
 * Lägen:
 *   node scripts/db/generation-history.mjs --json [--limit=100]   # senaste genereringar (lista)
 *   node scripts/db/generation-history.mjs --json --chat=<chatId> # drilldown för en chatt
 *
 * Körs read-only mot DB:n som `.env.local` pekar på. Mönster speglar
 * `scripts/db/scaffold-scores.mjs` (ingen Python-DB-driver i backoffice).
 */
import { config } from "dotenv";
import pg from "pg";
import { normalizeEnvUrl, warnIfProdLikeReadTarget } from "./db-target-guard.mjs";

config({ path: ".env.local" });
warnIfProdLikeReadTarget({ commandName: "generation-history" });

const argv = process.argv.slice(2);
const wantJson = argv.includes("--json");
const allowInsecureSsl = argv.includes("--allow-insecure-ssl");

function argValue(name) {
  const inline = argv.find((a) => a.startsWith(`--${name}=`));
  if (inline) return inline.slice(name.length + 3);
  const idx = argv.indexOf(`--${name}`);
  if (idx !== -1 && idx + 1 < argv.length && !argv[idx + 1].startsWith("--")) {
    return argv[idx + 1];
  }
  return null;
}

const chatId = argValue("chat");
const limitRaw = Number.parseInt(argValue("limit") ?? "100", 10);
const limit = Number.isFinite(limitRaw) ? Math.min(Math.max(limitRaw, 1), 500) : 100;

function emitError(message) {
  if (wantJson) {
    process.stdout.write(JSON.stringify({ error: message }));
  } else {
    console.error(message);
  }
}

const cs = normalizeEnvUrl(
  process.env.POSTGRES_URL ||
    process.env.POSTGRES_URL_NON_POOLING ||
    process.env.STORAGE_POSTGRES_URL ||
    process.env.STORAGE_POSTGRES_URL_NON_POOLING ||
    process.env.DATABASE_URL,
);
if (!cs) {
  emitError("Database URL missing (.env.local / pulled env).");
  process.exit(1);
}

const url = new URL(cs);
url.searchParams.delete("sslmode");
url.searchParams.delete("supa");

function resolveSsl() {
  const raw = process.env.DB_SSL_REJECT_UNAUTHORIZED?.trim().toLowerCase();
  if (raw === "false" || allowInsecureSsl) {
    return { rejectUnauthorized: false };
  }
  return { rejectUnauthorized: true };
}

const client = new pg.Client({ connectionString: url.toString(), ssl: resolveSsl() });

/**
 * Project verifier/autofix outcome fields from telemetry `meta` onto each row.
 * Older rows lack these keys — keep null rather than inventing defaults.
 */
function projectTelemetryMetaFields(row) {
  const meta = row?.meta && typeof row.meta === "object" && !Array.isArray(row.meta) ? row.meta : null;
  const verifier =
    meta?.postStreamSteps &&
    typeof meta.postStreamSteps === "object" &&
    !Array.isArray(meta.postStreamSteps)
      ? meta.postStreamSteps.verifier
      : null;
  const autofix = meta?.autofix && typeof meta.autofix === "object" && !Array.isArray(meta.autofix)
    ? meta.autofix
    : null;
  const verifierObj =
    verifier && typeof verifier === "object" && !Array.isArray(verifier) ? verifier : null;
  return {
    ...row,
    verifier_findings_before:
      typeof verifierObj?.findingsBefore === "number" ? verifierObj.findingsBefore : null,
    verifier_findings_after:
      typeof verifierObj?.findingsAfter === "number" ? verifierObj.findingsAfter : null,
    verifier_fixer_used:
      typeof verifierObj?.fixerUsed === "boolean" ? verifierObj.fixerUsed : null,
    verifier_fixer_improved:
      typeof verifierObj?.fixerImproved === "boolean" ? verifierObj.fixerImproved : null,
    autofix_risky_count:
      typeof autofix?.riskyFixCount === "number" ? autofix.riskyFixCount : null,
  };
}

/**
 * Keep in sync with `resolveReportedQualityGateResult` in
 * `src/lib/db/services/reported-quality-gate.ts`. The Node script cannot
 * import that TS module; the rule is four lines and the vitest lock lives
 * next to the TypeScript owner.
 */
function resolveReportedQualityGateResult(qualityGateResult, productBlocked) {
  if (qualityGateResult === "preflight_passed" && productBlocked === true) {
    return "product_blocked";
  }
  return qualityGateResult ?? null;
}

function projectHistoryRow(row) {
  const projected = projectTelemetryMetaFields(row);
  projected.reported_quality_gate = resolveReportedQualityGateResult(
    projected.quality_gate_result,
    projected.product_blocked === true,
  );
  return projected;
}

const LATEST_PRODUCT_POSTCHECK_JOIN = `
  LEFT JOIN LATERAL (
    SELECT (e.meta @> '{"productBlocked": true}'::jsonb) AS product_blocked
    FROM engine_version_error_logs e
    WHERE e.version_id = gt.version_id
      AND e.category = 'product_postcheck.summary'
    ORDER BY e.created_at DESC
    LIMIT 1
  ) pps ON true
`;

const RECENT_QUERY = `
  SELECT
    gt.created_at,
    gt.chat_id,
    gt.version_id,
    gt.scaffold_id,
    gt.model,
    gt.model_tier,
    gt.build_intent,
    gt.build_method,
    gt.prompt_classification,
    gt.retry_count,
    gt.autofix_applied,
    gt.syntax_fixer_used,
    gt.preflight_error_count,
    gt.preflight_warning_count,
    gt.quality_gate_result,
    pps.product_blocked,
    gt.preview_success,
    gt.preview_blocking_reason,
    gt.deploy_result,
    gt.duration_ms,
    gt.file_count,
    gt.meta,
    v.version_number,
    v.lifecycle_stage,
    v.release_state,
    v.verification_state,
    c.title        AS chat_title,
    c.project_id,
    p.name         AS project_name
  FROM generation_telemetry gt
  ${LATEST_PRODUCT_POSTCHECK_JOIN.trim()}
  LEFT JOIN engine_versions v ON v.id = gt.version_id
  LEFT JOIN engine_chats   c ON c.id = gt.chat_id
  LEFT JOIN app_projects   p ON p.id = c.project_id
  ORDER BY gt.created_at DESC
  LIMIT $1
`;

const CHAT_META_QUERY = `
  SELECT c.id, c.title, c.project_id, c.scaffold_id, c.created_at, p.name AS project_name
  FROM engine_chats c
  LEFT JOIN app_projects p ON p.id = c.project_id
  WHERE c.id = $1
`;

const CHAT_VERSIONS_QUERY = `
  SELECT id, version_number, lifecycle_stage, release_state, verification_state,
         verification_summary, preview_url, parent_version_id, created_at
  FROM engine_versions
  WHERE chat_id = $1
  ORDER BY version_number ASC
`;

const CHAT_TELEMETRY_QUERY = `
  SELECT gt.version_id, gt.scaffold_id, gt.model, gt.model_tier, gt.build_intent, gt.build_method,
         gt.prompt_classification, gt.retry_count, gt.autofix_applied, gt.syntax_fixer_used,
         gt.preflight_error_count, gt.preflight_warning_count, gt.quality_gate_result,
         pps.product_blocked,
         gt.preview_success, gt.preview_blocking_reason, gt.deploy_result, gt.duration_ms,
         gt.file_count, gt.meta, gt.created_at
  FROM generation_telemetry gt
  ${LATEST_PRODUCT_POSTCHECK_JOIN.trim()}
  WHERE gt.chat_id = $1
  ORDER BY gt.created_at ASC
`;

const CHAT_ERROR_LOGS_QUERY = `
  SELECT version_id, level, category, message, created_at
  FROM engine_version_error_logs
  WHERE chat_id = $1
  ORDER BY created_at ASC
  LIMIT 500
`;

const CHAT_GEN_LOGS_QUERY = `
  SELECT model, prompt_tokens, completion_tokens, duration_ms, success, error_message, created_at
  FROM engine_generation_logs
  WHERE chat_id = $1
  ORDER BY created_at ASC
  LIMIT 200
`;

try {
  await client.connect();

  if (chatId) {
    const [meta, versions, telemetry, errorLogs, genLogs] = await Promise.all([
      client.query(CHAT_META_QUERY, [chatId]),
      client.query(CHAT_VERSIONS_QUERY, [chatId]),
      client.query(CHAT_TELEMETRY_QUERY, [chatId]),
      client.query(CHAT_ERROR_LOGS_QUERY, [chatId]),
      client.query(CHAT_GEN_LOGS_QUERY, [chatId]),
    ]);
    const payload = {
      mode: "chat",
      generatedAt: new Date().toISOString(),
      chat: meta.rows[0] ?? { id: chatId, missing: true },
      versions: versions.rows,
      telemetry: telemetry.rows.map(projectHistoryRow),
      errorLogs: errorLogs.rows,
      generationLogs: genLogs.rows,
    };
    if (wantJson) {
      process.stdout.write(JSON.stringify(payload));
    } else {
      console.log(`Chat ${chatId}: ${versions.rows.length} versioner, ${errorLogs.rows.length} fel-rader.`);
    }
    process.exit(0);
  }

  const result = await client.query(RECENT_QUERY, [limit]);
  const rows = result.rows.map(projectHistoryRow);
  let success = 0;
  let failed = 0;
  let pending = 0;
  for (const r of rows) {
    if (r.preview_success === true) success += 1;
    else if (r.preview_success === false) failed += 1;
    else pending += 1;
  }
  const payload = {
    mode: "recent",
    limit,
    generatedAt: new Date().toISOString(),
    summary: { total: rows.length, success, failed, pending },
    rows,
  };
  if (wantJson) {
    process.stdout.write(JSON.stringify(payload));
  } else {
    console.log(`Senaste ${rows.length} genereringar (success=${success}, failed=${failed}, pending=${pending}).`);
    for (const r of rows.slice(0, 30)) {
      console.log(
        [
          (r.created_at instanceof Date ? r.created_at.toISOString() : String(r.created_at)).slice(0, 19),
          r.project_name || "—",
          r.scaffold_id || "—",
          r.model || "—",
          r.preview_success === true ? "ok" : r.preview_success === false ? "FAIL" : "pending",
        ].join("\t"),
      );
    }
  }
  process.exit(0);
} catch (err) {
  emitError(err instanceof Error ? err.message : String(err));
  process.exit(1);
} finally {
  await client.end().catch(() => {});
}

File: backoffice/pages/generation_history.py
"""Generation History — tidigare genereringar av användarsidor + vad varje steg gjorde.

Read-only mot DB:n som `.env.local` pekar på. Kör Node-scriptet
`scripts/db/generation-history.mjs --json` via subprocess (samma mönster som
`scaffold_performance.py` — backoffice har ingen egen Postgres-driver) och
renderar resultatet.

Två lager:
  1. Senaste genereringar (lista) — en rad per `generation_telemetry`-post med
     scaffold, model, retry/Normalize, preflight, RenderGate/ReleaseGate och preview-utfall.
  2. Drilldown per chatt — versioner (F2/F3 lifecycle), telemetri, fel-loggar
     (`engine_version_error_logs`) och generation-loggar.

Datakällor (skrivs av pipelinen): `generation_telemetry`, `engine_versions`,
`engine_version_error_logs`, `engine_generation_logs`, `engine_chats`,
`app_projects`. Inget skrivs härifrån.
"""

from __future__ import annotations

import json
import subprocess
from typing import Any

import pandas as pd
import streamlit as st

from backoffice.shared import BackofficeContext

_SCRIPT_REL = "scripts/db/generation-history.mjs"
_TIMEOUT_S = 60


def _run_history(repo_root, extra_args: list[str]) -> dict[str, Any]:
    """Kör generation-history.mjs read-only och returnerar parsad JSON.

    Returnerar alltid en dict; vid fel ligger orsaken i nyckeln ``error``.
    """
    script_path = repo_root / _SCRIPT_REL
    if not script_path.exists():
        return {"error": f"Script saknas: {script_path}"}
    try:
        result = subprocess.run(
            ["node", str(script_path), "--json", *extra_args],
            cwd=str(repo_root),
            capture_output=True,
            text=True,
            timeout=_TIMEOUT_S,
            check=False,
        )
    except subprocess.TimeoutExpired:
        return {"error": f"Script timeout efter {_TIMEOUT_S}s"}
    except FileNotFoundError:
        return {"error": "`node` saknas på PATH"}

    stdout = (result.stdout or "").strip()
    if not stdout:
        return {"error": (result.stderr or "Tomt svar från script").strip()}
    try:
        data = json.loads(stdout)
    except json.JSONDecodeError as exc:
        return {"error": f"Kunde inte tolka JSON: {exc}"}
    if isinstance(data, dict):
        return data
    return {"error": "Oväntat svarsformat (förväntade objekt)."}


def _preview_label(value: Any) -> str:
    # M#pv1 (honest preview_success tri-state):
    #   True  -> runtime confirmed ready (runtime-ready receipt), not just
    #            "preflight did not block" as the old semantics claimed.
    #   False -> confirmed no working preview (blocked, or session start failed).
    #   None  -> pending / unconfirmed (fresh boot queued, or no preview attempt).
    # Rows before the 2026-07 semantic cutoff used the old over-optimistic
    # meaning, so an older "ready" is weaker evidence than a current one.
    if value is True:
        return "ready"
    if value is False:
        return "failed"
    return "pending"


def _short(value: Any, length: int = 36) -> str:
    text = "" if value is None else str(value)
    return text[:length]


def _recent_dataframe(rows: list[dict[str, Any]]) -> pd.DataFrame:
    out = []
    for r in rows:
        out.append(
            {
                "Tid": _short(r.get("created_at"), 19),
                "Projekt": r.get("project_name") or "—",
                "Chatt": r.get("chat_title") or _short(r.get("chat_id"), 12),
                "v#": r.get("version_number"),
                "Stage": r.get("lifecycle_stage") or "—",
                "Scaffold": r.get("scaffold_id") or "—",
                "Model": r.get("model") or "—",
                "Intent": r.get("build_intent") or "—",
                "Preview": _preview_label(r.get("preview_success")),
                "Quality gate": r.get("quality_gate_result") or "—",
                "Deploy": r.get("deploy_result") or "—",
                "Retry": r.get("retry_count"),
                "Autofix": "ja" if r.get("autofix_applied") else "nej",
                "Preflight E/W": f"{r.get('preflight_error_count', 0)}/{r.get('preflight_warning_count', 0)}",
                "Blocking": _short(r.get("preview_blocking_reason"), 40) or "—",
                "Filer": r.get("file_count"),
                "chat_id": r.get("chat_id") or "",
            }
        )
    return pd.DataFrame(out)


def _render_chat_detail(ctx: BackofficeContext, chat_id: str) -> None:
    payload = _run_history(ctx.repo_root, [f"--chat={chat_id}"])
    if payload.get("error"):
        st.error(f"Kunde inte hämta chatt-detaljer: {payload['error']}")
        return

    chat = payload.get("chat") or {}
    if chat.get("missing"):
        st.warning(f"Hittade ingen chatt med id `{chat_id}`.")
        return

    st.markdown(
        f"**Projekt:** {chat.get('project_name') or '—'} · "
        f"**Chatt:** {chat.get('title') or '—'} · "
        f"`{chat.get('id', chat_id)}`"
    )

    versions = payload.get("versions") or []
    st.markdown(f"#### Versioner ({len(versions)})")
    if versions:
        st.dataframe(
            pd.DataFrame(
                [
                    {
                        "v#": v.get("version_number"),
                        "Stage": v.get("lifecycle_stage") or "—",
                        "Release": v.get("release_state") or "—",
                        "Verify": v.get("verification_state") or "—",
                        "Parent": _short(v.get("parent_version_id"), 12) or "—",
                        "Preview-URL": "ja" if v.get("preview_url") else "nej",
                        "Tid": _short(v.get("created_at"), 19),
                        "version_id": _short(v.get("id"), 16),
                    }
                    for v in versions
                ]
            ),
            hide_index=True,
            use_container_width=True,
        )
    else:
        st.caption("Inga versioner — chatten kan ha aborterat innan en version skapades.")

    telemetry = payload.get("telemetry") or []
    if telemetry:
        st.markdown(f"#### Telemetri per generering ({len(telemetry)})")
        st.dataframe(
            pd.DataFrame(
                [
                    {
                        "Tid": _short(t.get("created_at"), 19),
                        "Scaffold": t.get("scaffold_id") or "—",
                        "Model": t.get("model") or "—",
                        "Intent": t.get("build_intent") or "—",
                        "Preview": _preview_label(t.get("preview_success")),
                        "Quality gate": t.get("quality_gate_result") or "—",
                        "Deploy": t.get("deploy_result") or "—",
                        "Retry": t.get("retry_count"),
                        "Autofix": "ja" if t.get("autofix_applied") else "nej",
                        "Syntax-fixer": "ja" if t.get("syntax_fixer_used") else "nej",
                        "Preflight E/W": f"{t.get('preflight_error_count', 0)}/{t.get('preflight_warning_count', 0)}",
                        "Blocking": _short(t.get("preview_blocking_reason"), 40) or "—",
                    }
                    for t in telemetry
                ]
            ),
            hide_index=True,
            use_container_width=True,
        )

    error_logs = payload.get("errorLogs") or []
    st.markdown(f"#### Fel/warnings ({len(error_logs)})")
    if error_logs:
        st.dataframe(
            pd.DataFrame(
                [
                    {
                        "Tid": _short(e.get("created_at"), 19),
                        "Level": e.get("level") or "—",
                        "Kategori": e.get("category") or "—",
                        "Version": _short(e.get("version_id"), 12),
                        "Meddelande": _short(e.get("message"), 120),
                    }
                    for e in error_logs
                ]
            ),
            hide_index=True,
            use_container_width=True,
        )
    else:
        st.caption("Inga fel-/warning-rader för den här chatten.")

    gen_logs = payload.get("generationLogs") or []
    if gen_logs:
        st.markdown(f"#### Generation-loggar ({len(gen_logs)})")
        st.dataframe(
            pd.DataFrame(
                [
                    {
                        "Tid": _short(g.get("created_at"), 19),
                        "Model": g.get("model") or "—",
                        "Prompt tokens": g.get("prompt_tokens"),
                        "Completion tokens": g.get("completion_tokens"),
                        "Duration ms": g.get("duration_ms"),
                        "Success": "ja" if g.get("success") else "nej",
                        "Fel": _short(g.get("error_message"), 80) or "—",
                    }
                    for g in gen_logs
                ]
            ),
            hide_index=True,
            use_container_width=True,
        )


def render(ctx: BackofficeContext) -> None:
    st.title("Generation History")
    st.caption(
        "Tidigare genereringar av användarsidor + vad varje steg gjorde "
        "(scaffold, model, retry/Normalize, preflight, RenderGate/ReleaseGate, preview-utfall). "
        "Read-only mot DB:n som `.env.local` pekar på. Källor: `generation_telemetry`, "
        "`engine_versions`, `engine_version_error_logs`, `engine_generation_logs`."
    )

    limit = st.number_input(
        "Antal senaste genereringar",
        min_value=10,
        max_value=500,
        value=50,
        step=10,
        help="Hur många rader (nyast först) som hämtas från generation_telemetry.",
    )

    with st.spinner("Hämtar genererings-historik ..."):
        payload = _run_history(ctx.repo_root, [f"--limit={int(limit)}"])

    if payload.get("error"):
        st.error(f"Kunde inte hämta data: {payload['error']}")
        st.caption(
            "Tips: kontrollera att `.env.local` har `POSTGRES_URL` (eller kör "
            "`npm run env:pull`) och att `node scripts/db/generation-history.mjs` kan köras lokalt."
        )
        return

    summary = payload.get("summary") or {}
    rows = payload.get("rows") or []
    col1, col2, col3, col4 = st.columns(4)
    col1.metric("Genereringar", summary.get("total", len(rows)))
    col2.metric("Preview ok", summary.get("success", "?"))
    col3.metric("Preview FAIL", summary.get("failed", "?"))
    col4.metric("Pending (ingen signal)", summary.get("pending", "?"))

    if not rows:
        st.info(
            "Ingen telemetri hittad. Antingen en färsk DB, eller så har inga "
            "genereringar körts ännu mot den DB som `.env.local` pekar på."
        )
        return

    st.subheader("Senaste genereringar")
    df = _recent_dataframe(rows)
    st.dataframe(df.drop(columns=["chat_id"]), hide_index=True, use_container_width=True)

    st.divider()
    st.subheader("Detaljer per chatt")
    options: dict[str, str] = {}
    for r in rows:
        cid = r.get("chat_id")
        if not cid or cid in options.values():
            continue
        label = (
            f"{r.get('project_name') or '—'} · "
            f"{r.get('chat_title') or _short(cid, 12)} · "
            f"{_short(r.get('created_at'), 19)}"
        )
        options[label] = cid

    if not options:
        st.caption("Inga chattar att visa detaljer för.")
        return

    selected_label = st.selectbox("Välj chatt", list(options.keys()), key="gen_history_chat_pick")
    if st.button("Visa detaljer", key="gen_history_chat_show"):
        st.session_state["gen_history_selected_chat"] = options[selected_label]

    selected_chat = st.session_state.get("gen_history_selected_chat")
    if selected_chat:
        st.divider()
        _render_chat_detail(ctx, selected_chat)
